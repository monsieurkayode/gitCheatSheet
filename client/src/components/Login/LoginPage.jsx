import React, { Fragment, Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { loginUser } from '../../actions/userActions';
import Container from '../Common/Container';
import Button from '../Common/Button';
import FormField from '../Common/FormField';
import Loader from '../Common/Loader';

class LoginPage extends Component {
  static propTypes = {
    loginUser: PropTypes.func.isRequired,
    authenticated: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired
  }

  static initialState = () => ({
    user: {
      username: '',
      password: ''
    },
    errors: {}
  })

  state = { ...LoginPage.initialState() }

  handleInputChange = (event) => {
    const { user, errors } = this.state;
    user[event.target.name] = event.target.value;
    errors[event.target.name] = '';
    this.setState({ user, errors });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const errors = this.validateUser(this.state.user);

    if (Object.keys(errors).length) {
      return this.setState({ errors });
    }

    this.props.loginUser(this.state.user);
  }

  handleBlur = (event) => {
    const { errors } = this.state;
    errors[event.target.name] = this.validateUser(
      this.state.user
    )[event.target.name];
    this.setState({ errors });
  }

  validateUser = ({ username, password }) => {
    const errors = {};

    if (!username.trim()) {
      errors.username = 'Enter your username';
    }

    if (!password) {
      errors.password = 'Enter your password to login';
    }

    return errors;
  }

  render() {
    const { user, errors } = this.state;
    const { authenticated, loading } = this.props;

    if (authenticated) {
      return <Redirect to="/" />;
    }

    return (
      <Fragment>
        <Container>
          <form onSubmit={this.handleSubmit}>
            <FormField
              icon="account_circle"
              placeholder="Username"
              name="username"
              value={user.username}
              error={errors.username}
              onChange={this.handleInputChange}
              onBlur={this.handleBlur}
              loading={loading}
            />
            <FormField
              icon="lock_open"
              type="password"
              placeholder="Password"
              name="password"
              value={user.password}
              error={errors.password}
              onChange={this.handleInputChange}
              onBlur={this.handleBlur}
              loading={loading}
            />
            <div className="form-field">
              <Button type="submit" value="Login" disabled={loading} />
              <Loader size={15} strokeColor="#fff" inline loading={loading} />
            </div>
          </form>
        </Container>
      </Fragment>
    );
  }
}

const mapStateToProps = ({ userDetails }) => ({
  authenticated: Object.keys(userDetails.user).length === 2,
  loading: userDetails.makingApiRequest
});

export default connect(mapStateToProps, { loginUser })(LoginPage);
