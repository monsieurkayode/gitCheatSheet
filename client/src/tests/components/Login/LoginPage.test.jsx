import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage, mapStateToProps } from '../../../components/Login/LoginPage';

const setup = () => {
  const props = {
    authenticated: false,
    loading: false,
    loginUser: () => Promise.resolve()
  };

  return shallow(<LoginPage {...props} />);
};

describe('LoginPage Component', () => {
  const wrapper = setup();
  const event = { target: { name: 'username', value: '' }, preventDefault: jest.fn() };

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleInputChange', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleInputChange');
    wrapper.instance().handleInputChange(event);
    expect(spy).toHaveBeenCalled();
  });

  describe('handleSubmit', () => {
    it('should be called', () => {
      const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');
      wrapper.instance().handleSubmit(event);
      expect(spy).toHaveBeenCalled();
    });

    it('should set error if form is invalid', () => {
      wrapper.instance().handleSubmit(event);
      expect(wrapper.state('errors'))
        .toEqual({
          password: 'Enter your password to login',
          username: 'Enter your username'
        });
    });

    it('should submit valid forms with no errors', () => {
      wrapper.setState({ user: { username: 'user', password: 'password' }, errors: {} });
      wrapper.instance().handleSubmit(event);
      expect(wrapper.state('errors')).toEqual({});
    });
  });


  it('should call handleBlur', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleBlur');
    wrapper.instance().handleBlur(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should return a Redirect component if user is authenticated', () => {
    wrapper.setProps({ authenticated: true });
    expect(wrapper.find('Redirect').length).toBe(1);
  });

  describe('mapStateToProps', () => {
    it('should return an object', () => {
      expect(mapStateToProps({ userDetails: { user: {}, makingApiRequest: true } }))
        .toBeInstanceOf(Object);
    });
  });
});
