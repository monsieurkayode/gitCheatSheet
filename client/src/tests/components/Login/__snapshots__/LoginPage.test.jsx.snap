// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Component should render correctly 1`] = `
<Fragment>
  <styled.main>
    <form
      onSubmit={[Function]}
    >
      <FormField
        error=""
        icon="account_circle"
        loading={false}
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <FormField
        error=""
        icon="lock_open"
        loading={false}
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        className="form-field"
      >
        <styled.input
          disabled={false}
          type="submit"
          value="Login"
        />
        <Loader
          inline={true}
          loading={false}
          size={15}
          strokeColor="#fff"
        />
      </div>
    </form>
  </styled.main>
</Fragment>
`;
